crates/astria-bridge-withdrawer/src/bridge_withdrawer/submitter/frost_signer.rs

```diff
@@ -0,0 +1,320 @@
+
+        ensure!(
+            self.participant_clients.len() == min_signers,
+            "not enough participant clients; need at least {min_signers}"
+        );
+
+        Ok(FrostSigner {
+            min_signers,
+            public_key_package,
+            address,
+            participant_clients: self.participant_clients,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct FrostSigner {
+    min_signers: usize,
+    public_key_package: PublicKeyPackage,
+    address: Address,
+    participant_clients:
+        HashMap<Identifier, FrostParticipantServiceClient<tonic::transport::Channel>>,
+}
+
+#[tonic::async_trait]
+impl Signer for FrostSigner {
+    async fn sign(&self, tx: TransactionBody) -> eyre::Result<Transaction> {
+        // part 1: gather commitments from participants
+        let (commitments, signing_package_commitments) = self.frost_part_1().await;
+        ensure!(
+            commitments.len() >= self.min_signers,
+            "not enough part 1 commitments received; want at least {}, got {}",
+            self.min_signers,
+            commitments.len()
+        );
+
+        // part 2: gather signature shares from participants
+        let tx_bytes = tx.to_raw().encode_to_vec();
+        let sig_shares = self.frost_part_2(commitments, tx_bytes.clone()).await;
+        ensure!(
+            sig_shares.len() >= self.min_signers,
+            "not enough part 2 signature shares received; want at least {}, got {}",
+            self.min_signers,
+            sig_shares.len()
+        );
+
+        // finally, aggregate and create signature
+        let signing_package =
+            frost_ed25519::SigningPackage::new(signing_package_commitments, &tx_bytes);
+        let signature =
+            frost_ed25519::aggregate(&signing_package, &sig_shares, &self.public_key_package)
+                .wrap_err("failed to aggregate signature shares")?;
+
+        let raw_transaction = astria_core::generated::astria::protocol::transaction::v1::Transaction {
+                body: Some(pbjson_types::Any {
+                    type_url: astria_core::generated::astria::protocol::transaction::v1::TransactionBody::type_url(),
+                    value: tx_bytes.into(),
+                }),
+                signature: signature
+                    .serialize()
+                    .wrap_err("failed to serialize signature")?
+                    .into(),
+                public_key: self.public_key_package
+                    .verifying_key()
+                    .serialize()
+                    .wrap_err("failed to serialize verifying key")?
+                    .into(),
+            };
+        let transaction = Transaction::try_from_raw(raw_transaction)
+            .wrap_err("failed to convert raw transaction to transaction")?;
+
+        Ok(transaction)
+    }
+
+    fn address(&self) -> &Address {
+        &self.address
+    }
+}
+
+impl FrostSigner {
+    async fn frost_part_1(
```

Comment:
Is there a verb we can use here than `forst_part_1`? Maybe `perform_part_1`? Or `execute_part_1`?